### PR TITLE
Add `J.UnaryTypeTree` for Kotlin

### PR DESCRIPTION
**Pull request by kunli2**
Languages like Kotlin support a unary type modifier such as `?` after the type, indicating nullable.
Adding `J.UnaryTypeTree` to support those cases as a general fix.
The according change for kotlin : https://github.com/openrewrite/rewrite-kotlin/pull/497

### CODE REVIEW

**Comment by knutwannheden** [comment]
@kunli2 Are we already aware of any other uses that this may have?

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 1]
```diff
@@ -2389,6 +2389,109 @@ public CoordinateBuilder.Expression getCoordinates() {
...
+        public enum Type {
+            IsNullable;
+
+            public boolean isModifying() {
+                switch (this) {
+                    case IsNullable:
+                    default:
+                        return false;
+                }
+            }
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2458]: What is this for?

Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2458]: Ah, this should be removed.

**Comment by kunli2** [comment]
> @kunli2 Are we already aware of any other uses that this may have?

I think for use site annotation, currently we use the marker `AnnotationUseSite`, 
an example of use site is like 
```kotlin
@field /*C1*/  :  [   Annotation1    Annotation2 (  "some parameters"   )    ]
```
Currently we use `AnnotationUseSite` to store the space before the colon `/*C1*/`, and we do want to remove that marker.
Which I think maybe we can use J.UnaryTypeTree for `@field /*C1*/ :`.

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 2]
```diff
@@ -2389,6 +2389,109 @@ public CoordinateBuilder.Expression getCoordinates() {
...
 
+    /**
+     * Languages like Kotlin support a unary type modifier such as `?` after the type, indicating nullable.
+     */
+    @FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+    @EqualsAndHashCode(callSuper = false, onlyExplicitlyIncluded = true)
+    @RequiredArgsConstructor
+    @AllArgsConstructor(access = AccessLevel.PRIVATE)
+    @With
+    class UnaryTypeTree implements J, TypeTree, Expression {
```
[the remark this answers is not in the document]
Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2400]: good point! Sam.
making it a `NullableTypeTree` makes sense

**Comment by traceyyoshima** [comment]
> > @kunli2 Are we already aware of any other uses that this may have?
> 
> I think for use site annotation, currently we use the marker `AnnotationUseSite`, an example of use site is like
> 
> ```kotlin
> @field /*C1*/  :  [   Annotation1    Annotation2 (  "some parameters"   )    ]
> ```
> 
> Currently we use `AnnotationUseSite` to store the space before the colon `/*C1*/`, and we do want to remove that marker. `@field` is a `NameTree`, I think maybe we can use `J.UnaryTypeTree` to present `@field /*C1*/ :` with a colon like unary, but I am not sure.

Other markers we'll be replacing:
- IsNullable:
    - type of field: TypeTree
- TypeReferencePrefix:
    - type of field: TypeTree

- AnnotationUseSite:
    - type of field: NameTree (TypeTree extends NameTree)

- TrailingComma:
    - type of field: requires investigation

- EmptyArgumentListPrecedesArgument
    - type of field: requires investigation

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java (hunk 1 above)
Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java]: removed.

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (traceyyoshima) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2400]: We could technically have a `NullableTypeTree` and a class like a `PaddedTypeTree` to handle the rest of the cases to replace TrailingComma and TypeReferences. Both trailing commas and types that follow `:` are just whitespace. Whereas a NullableTypeTree represents something meaningful in terms of the LST.

[the remark this answers is not in the document]
Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2400]: I think `PaddedTypeTree` is a more generic way and it makes sense to me.

[the remark this answers is not in the document]
Comment (kunli2) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2400]: @knutwannheden what do you think about this?

[the remark this answers is not in the document]
Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:2400]: For trailing commas I think we should stay with a marker, even if it owns a space. The reason being that neither the space nor the comma are required, so adding an element to the LST model would more likely just cause unnecessary confusion.

I would have to take another look at TypeReferences. Possibly we should also there have a more "dedicated" LST element.

**Pull request closed without merging**